pkg/console/controllers/clidownloads/controller.go

```diff
@@ -0,0 +1,261 @@
+	if err != nil {
+		return err
+	}
+	updatedOperatorConfig := operatorConfig.DeepCopy()
+
+	consoleRoute, err := c.routeClient.Routes(api.TargetNamespace).Get(api.OpenShiftConsoleRouteName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	host := routesub.GetCanonicalHost(consoleRoute)
+	ocConsoleCLIDownloads := PlatformBasedOCConsoleCLIDownloads(host, "amd64", ocCLIDownloadsCustomResourceName)
+	_, ocCLIDownloadsErrReason, ocCLIDownloadsErr := CreateOrUpdate(c.consoleCliDownloadsClient, ocConsoleCLIDownloads)
+	status.HandleDegraded(updatedOperatorConfig, "ConsoleCliDownloadsSync", ocCLIDownloadsErrReason, ocCLIDownloadsErr)
+	if ocCLIDownloadsErr != nil {
+		return ocCLIDownloadsErr
+	}
+
+	_, odoCLIDownloadsErrReason, odoCLIDownloadsErr := CreateOrUpdate(c.consoleCliDownloadsClient, ODOConsoleCLIDownloads())
+	status.HandleDegraded(updatedOperatorConfig, "ConsoleCliDownloadsSync", odoCLIDownloadsErrReason, odoCLIDownloadsErr)
+	if odoCLIDownloadsErr != nil {
+		return odoCLIDownloadsErr
+	}
+
+	return nil
+}
+
+func GetPlatformURL(baseURL, platform, archiveType string) string {
+	return fmt.Sprintf("%s/%s/%s", baseURL, platform, archiveType)
+}
+
+func PlatformBasedOCConsoleCLIDownloads(host, arch, cliDownloadsName string) *v1.ConsoleCLIDownload {
+	baseURL := fmt.Sprintf("%s/%s", util.HTTPS(host), arch)
+
+	return &v1.ConsoleCLIDownload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: cliDownloadsName,
+		},
+		Spec: v1.ConsoleCLIDownloadSpec{
+			Description: `With the OpenShift command line interface, you can create applications and manage OpenShift projects from a terminal.
+
+The oc binary offers the same capabilities as the kubectl binary, but it is further extended to natively support OpenShift Container Platform features.
+`,
+			DisplayName: "oc - OpenShift Command Line Interface (CLI)",
+			Links: []v1.Link{
+				{
+					Href: GetPlatformURL(baseURL, LinuxPlatformKey, "oc.tar"),
+					Text: "Download oc for Linux",
+				},
+				{
+					Href: GetPlatformURL(baseURL, MacPlatformKey, "oc.zip"),
+					Text: "Download oc for Mac",
+				},
+				{
+					Href: GetPlatformURL(baseURL, WindowsPlatformKey, "oc.zip"),
+					Text: "Download oc for Windows",
+				},
+			},
+		},
+	}
+}
+
+func ODOConsoleCLIDownloads() *v1.ConsoleCLIDownload {
+	return &v1.ConsoleCLIDownload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: odoCLIDownloadsCustomResourceName,
+		},
+		Spec: v1.ConsoleCLIDownloadSpec{
+			Description: `OpenShift Do (odo) is a fast, iterative, and straightforward CLI tool for developers who write, build, and deploy applications on OpenShift.
+
+odo abstracts away complex Kubernetes and OpenShift concepts, thus allowing developers to focus on what is most important to them: code.
+`,
+			DisplayName: "odo - Developer-focused CLI for OpenShift",
+			Links: []v1.Link{
+				{
+					Href: "https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/",
+					Text: "Download odo",
+				},
+			},
+		},
+	}
+}
+
```

Comment:
Pondering if we should make this 

```golang
ApplyConsoleCLIDownload(....)
``` 
We can probably look at a func in `library-go` like:

```golang
func ApplyService(client coreclientv1.ServicesGetter, recorder events.Recorder, required *corev1.Service) (*corev1.Service, bool, error) {
  // does things
}
```
and copy it.  I imagine we want the `eventRecorder` in there to create events if things go awry. 

So something like:

```golang
func ApplyConsoleCLIDownload(cliDownloadsClient CliDownloadsGetter, recorder events.Recorder, required *v1.ConsoleCLIDownload) (v1.ConsoleCLIDownload, string, error) {
  // more or less copy the ApplyService func, but tweak it to fit this resource.
}
```

Reply:
`CreateOrUpdate` already has similar logic... need to add the recored and rename though :)

Reply:
so looked at adding the recorder but noticed that the methods that the recorder is passed into are private (`reportCreateEvent` && `reportUpdateEvent`)

Also our custom apply [CustomApplyOAuth](https://github.com/openshift/console-operator/blob/master/pkg/console/subresource/oauthclient/oauthclient.go#L21) method is not using the recored as well, supposedly due to the fact the methods are private.

Would ship it for now and add this as a techDebt ?

Reply:
Agree!

Do you want to throw a couple `TODO` comments in there quick?  This would be helpful if we promote it to `openshift/library-go/pkg/console/resource/resourceapply/core.go` (does not yet exist, but if others are going to use these in their operators, we could ensure the appropriate `apply` funcs exist for each of our custom resources.)